[MM-38891] - CRT: threads list loading indicator and search guidance …

julmondragon opened the pull request:
#### Summary
This PR adds two enhancements to the CRT feature for the Threads global list:

- Adds a loading indicator each time more posts are being fetched on infinite scroll.
- When no more posts can be loaded (due to the historical limit of threads), provides search guidance at the bottom of the list. This is only shown if you’ve scrolled to load older posts and then reach the end of the list.

#### Ticket Link
Issue: https://github.com/mattermost/mattermost-server/issues/18796
JIRA: https://mattermost.atlassian.net/browse/MM-38891

#### Screenshots

![MM-38891](https://user-images.githubusercontent.com/79058848/146955848-40752bfd-751e-4e1c-aa82-0b78d3fa5543.gif)

#### Release Note
```release-note
CRT - Threads global list: Added a loading indicator each time more posts are being fetched on infinite scroll.
```

```release-note
CRT - Threads global list: Added a search guidance when no more posts can be loaded. This is only shown if you’ve scrolled to load older posts and then reach the end of the list.
```

mattermod commented:
Hello @julmondragon,

Thanks for your pull request! A Core Committer will review your pull request soon. For code contributions, you can learn more about the review process [here](https://developers.mattermost.com/contribute/getting-started/code-review/).

julmondragon commented:
FYI @neallred
@calebroseland could you please help reviewing?

neallred commented on components/threading/global_threads/thread_list/virtualized_thread_list.tsx:
```diff
@@ -46,7 +50,15 @@ function VirtualizedThreadList({
+    const data = useMemo(
+        () => (
+            {
+                ids: hasLoaded && ids.length === total ? [...ids, Constants.THREADS_NO_RESULTS_ITEM_ID] : (isLoading && ids.length !== total && [...ids, Constants.THREADS_LOADING_INDICATOR_ITEM_ID]) || ids,
```
I saw an issue where the first attempt to load more wouldn't show the loading indicator, but subsequent loads would. I'm guessing its something related to the `isLoading` check here.

https://user-images.githubusercontent.com/13738432/146991231-26e44bac-7e9b-475f-9c08-c829990860a7.mp4

neallred reviewed (commented):
It looks really nice! The only issues I see are:
1) With some screen widths, it would be nice to have the no results indicator be bigger to accommodate all the content
2) For the first time the virtualized list is loading, it does not show the loading dots.

I left more details as inline comments.

matthewbirtch reviewed (changes requested):
Thanks for working on this @julmondragon. Beyond what @neallred mentioned about the responsiveness, there are some other styling adjustments needed as outlined below:

1. Change the padding on the container to `padding: 16px 16px 16px 24px;`
2. Change font-size on "That's the end of the list" header to: `font-size: 16px;` and `line-height: 24px`
3. Make the margin on the bottom of the header `margin-bottom: 0px;`
4. Need to re-export the SVG to the clipped size provided in the [figma file at this link](https://www.figma.com/file/B1FyrTYjrfoA9AMCqIQ9tA/%E2%9C%85MM-38891-Threads-Infinite-Scroll-Loader?node-id=233%3A25099). Once you've done that, you should have an SVG that is 72x48. Let me know if you need me to export this for you.
5. Change margin-right on the SVG to `margin-right: 16px`
5. Do we need to set an explicit height on the container?  Or should it grow to the height it needs to automatically? It looks like the height is explicitly set to 133px. If this needs to be an explicit height, this height will likely need to change based on the feedback above.
6. The shortcut style is quite different from the design. Quite a few changes need to be made to those to align with design:
```
font-size:12px;
font-weight: 600;
line-height: 16px;
padding: 2px 4px;
color: rgba(var(--center-channel-color-rgb),.72);
```

AshishDhama commented:
@julmondragon Thanks for working on this I think you are almost there, If you need some help on this one please let me know.

esethna commented:
Hey @julmondragon, thanks for the PR! How's this going? Are you open to finishing this up?

julmondragon commented:
Hi @esethna , yes! I'll resume the work on this ticket to address the feedback soon.

esethna commented:
Thanks @julmondragon!

julmondragon commented:
/e2e-test

julmondragon commented:
@neallred @matthewbirtch I think I addressed all the feedback received.

**Loading indicator and Search guidance**
![CRT-loading-search guidance](https://user-images.githubusercontent.com/79058848/154976051-83b78d86-0b38-41ca-b88f-bbbe09e44714.gif)


**Screen sizes**
![CRT-search guidance-dims](https://user-images.githubusercontent.com/79058848/154976211-1ac1fb8e-1624-40e7-9fe5-5befba66c0a9.gif)

matthewbirtch reviewed (approved):
Thanks @julmondragon! Looks good to me

pablovelezvidal commented:
/e2e-tes

pablovelezvidal reviewed (approved):
LGTM! great work @julmondragon

stevemudie reviewed (approved):
Looks good @julmondragon!

Scrolling worked as expected and the information on how to search at the bottom of the list was there as well.

esethna commented:
@julmondragon I think this is good to be merged, thanks for the work!
